src/backends/interflop-cancellation/interflop_cancellation.c

```diff
@@ -0,0 +1,280 @@
+ *                                                                           *
+ *  You should have received a copy of the GNU General Public License        *
+ *  along with Verificarlo.  If not, see <http://www.gnu.org/licenses/>.     *
+ *                                                                           *
+ *****************************************************************************/
+
+// Changelog:
+
+#include <argp.h>
+#include <err.h>
+#include <errno.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <sys/time.h>
+#include <unistd.h>
+#include <string.h>
+
+#include "../../common/float_const.h"
+#include "../../common/interflop.h"
+#include "../../common/tinymt64.h"
+
+#include "quadmath-imp.h"
+
+typedef struct {
+  int choose_seed;
+  uint64_t seed;
+} t_context;
+
+/* define default environment variables and default parameters */
+#define MCA_TOLERANCE_DEFAULT 1
+#define WARNING_DEFAULT 0
+
+static int WARN = WARNING_DEFAULT;
+
+static int MCALIB_T = MCA_TOLERANCE_DEFAULT;
+
+// possible op values
+#define MCA_ADD 1
+#define MCA_SUB 2
+#define MCA_MUL 3
+#define MCA_DIV 4
+
+#define min(a, b) ((a) < (b) ? (a) : (b))
+#define max(a, b) ((a) > (b) ? (a) : (b))
+#define bit(A,i) ((A >> i) & 1)
+#define bit_a_b(X,a,b) (((X << (sizeof(X)*8 - max(a,b) - 1)) >> (sizeof(X)*8 - max(a,b) - 1)) & ((X >> min(a,b) ) << min(a,b)))
+
+static float _mca_sbin(float a, float b, int qop);
+static double _mca_dbin(double a, double b, int qop);
+
+/******************** MCA CONTROL FUNCTIONS *******************
+ * The following functions are used to set virtual precision and
+ * MCA mode of operation.
+ ***************************************************************/
+
+static int _set_mca_tolerance(int tolerance) {
+  MCALIB_T = tolerance;
+  return 0;
+}
+
+static int _set_warning() {
+  WARN = 1;
+}
+
+/******************** MCA RANDOM FUNCTIONS ********************
+ * The following functions are used to calculate the random
+ * perturbations used for MCA
+ ***************************************************************/
+
+/* random generator internal state */
+static tinymt64_t random_state;
+
+static double _mca_rand(void) {
+  /* Returns a random double in the (0,1) open interval */
+  return tinymt64_generate_doubleOO(&random_state);
+}
+
+#define exp(X) ({                                     \
+  /* Convert X to an int */                           \
+  u_int64_t _V_ = *(u_int64_t*)&X;                    \
+  /* Get size of the exponent */                      \
+  int _E_ = (6+3*(sizeof(X) >> 2)-1);                 \ 
+  /* Get the size of the mantiss */                   \
+  int _M_ = (sizeof(X)*8-_E_-1);                      \
+  /* Compute the exponent */                          \     
+  (int)bit_a_b(_V_>>_M_,0,_E_-1) - ((1<<_E_-1)-2);})
+
+#define inexact(X,S) (X + pow(2,exp(X)-S) * _mca_rand())
```

Comment:
This can be optimized. Because noise is only added to cancellation bits (which are necessarily zero); it is enough to generate a rand field, shift it to the first cancelled bit and OR it to the result.

Reply:
Thinking about this. The advantage of using MCA noise is that it is centered on the IEEE result. The shift and or approach I propose will generate a bias... I'm not sure yet which model is better to simulate effect of cancellations on hardware. We could have both modes; but this requires further study.

Reply:
Actually after thinking about this, the more realistic noise is the one you have used (MCA noise).
Indeed cancellation can produce negative and positive errors, when doing a - b. If a is exact and b is imprecise. After cancellation we will have a negative term.

Therefore, we won't do the optimization I was suggesting above. Your first version is fine !

Reply:
I have changed your macros to use the new macros added by @yohanchatelain in #152. A small diffrence is that while mca backend uses a float128 to accumulate noise and rounds at the end. Here we do it directly in float64. But I think in the context of cancellation this is ok; since there is only one noise term. @yohanchatelain what do you think ?

Reply:
I personally don't get how work the macro like I mention [here](https://github.com/verificarlo/verificarlo/pull/153#discussion_r389141728) but I agree with you on the use of float64 instead of float128 in this context. @Avaquet ?